<?xml version="1.0"?>
<!--
Software License Agreement (BSD)

\file      description.launch
\authors   Paul Bovbel <pbovbel@clearpathrobotics.com>, Prasenjit Mukherjee <pmukherj@clearpathrobotics.com>
\copyright Copyright (c) 2015, Clearpath Robotics, Inc., All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that
the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the
   following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the 
   following disclaimer in the documentation and/or other materials provided with the distribution.
 * Neither the name of Clearpath Robotics nor the names of its contributors may be used to endorse or promote
   products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WAR-
RANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, IN-
DIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE
)ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->
<launch>

  <arg name="no_static_map" default="false"/>

  <arg name="base_global_planner" default="navfn/NavfnROS"/>
  <arg name="base_local_planner" default="mpc_local_planner/MpcLocalPlannerROS"/>
  <!-- <arg name="base_local_planner" default="base_local_planner/TrajectoryPlannerROS"/> ne radi dobro, ako je malo "zeznutija" orijentacija nece je zadovoljit a nece nista ni javit nego ce samo stat, ili ako mora blizu zidu proc ista stvar -->
  <arg name="cmd_vel_topic" default="cmd_vel" />
  <arg name="odom_topic" default="odom" />
  <!-- <arg name="node_name_prefix" default="husky1"/> -->
  <arg name="tf_prefix" default="husky1_tf"/>
  <arg name="scan_topic" default="/husky1/front/scan"/>

  <!-- <node pkg="move_base" type="move_base" respawn="false" name="$(arg node_name_prefix)_move_base" output="screen"> -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <remap from="map" to="map"/>
    <remap from="scan" to="$(arg scan_topic)"/>

    <param name="base_global_planner" value="$(arg base_global_planner)"/>
    <param name="base_local_planner" value="$(arg base_local_planner)"/>  
    <rosparam file="$(find husky_dynamic_navigation)/config/planner.yaml" command="load"/>
    <rosparam file="$(find husky_dynamic_navigation)/config/mpc_planner_params.yaml" command="load"/>

    <!-- observation sources located in costmap_common.yaml -->
    <rosparam file="$(find husky_dynamic_navigation)/config/costmap_common.yaml" command="load" ns="global_costmap" /> ispred svakog parama ima global_namespace/...
    <rosparam file="$(find husky_dynamic_navigation)/config/costmap_common.yaml" command="load" ns="local_costmap" />

    <!-- local costmap, needs size -->
    <rosparam file="$(find husky_dynamic_navigation)/config/costmap_local.yaml" command="load" ns="local_costmap" />
    <param name="local_costmap/width" value="5.0"/>
    <!-- <param name="local_costmap/width" value="3.0"/> sa ovim ne radi toliko problema kad trebaju doc u (0,2) i (0,-2) ali im se cesce mapa poremeti i lokalizacija,,, nakon sta sam mjenjao amcl i ovo dole za DWA radi super s ovim (ne zblesira se kad dode u (0,-3)),, tj nekad super radi a nekad ili husyk2 zapne (dode preblizu prepreci) ili mu se skroz mapa poremeti,, s ovim se isto dosta sporo gibaju -->
    <param name="local_costmap/height" value="5.0"/>
    <!-- <param name="local_costmap/height" value="3.0"/> -->

    <!-- static global costmap, static map provides size -->
    <rosparam file="$(find husky_dynamic_navigation)/config/costmap_global_static.yaml" command="load" ns="global_costmap" unless="$(arg no_static_map)"/>
    
    <!-- global costmap with laser, for odom_navigation_demo -->
    <rosparam file="$(find husky_dynamic_navigation)/config/costmap_global_laser.yaml" command="load" ns="global_costmap" if="$(arg no_static_map)"/>
    <param name="global_costmap/width" value="100.0" if="$(arg no_static_map)"/>
    <param name="global_costmap/height" value="100.0" if="$(arg no_static_map)"/>



    <!-- reset frame ids -->
    <param name="global_costmap/global_frame" value="map"/>
    <param name="global_costmap/robot_base_frame" value="$(arg tf_prefix)/base_link"/>
    <param name="local_costmap/global_frame" value="$(arg tf_prefix)/odom"/>
    <param name="local_costmap/robot_base_frame" value="$(arg tf_prefix)/base_link"/>
    <param name="MPCLocalPlannerROS/global_frame_id" value="$(arg tf_prefix)/odom"/>
    <param name="MPCLocalPlannerROS/odom_topic" value="$(arg tf_prefix)/husky_velocity_controller/odom"/>
    
    <!-- <param name="TrajectoryPlannerROS/global_frame_id" value="$(arg tf_prefix)/odom"/> ovisno dal koristis dwa ili ovog -->
    <!-- <param name="move_base/local_costmap/obstacles_laser/laser/topic" value="$(arg scan_topic)"/> nije potrebno stavljat move_base ispred imena parametra jer to doda sam node, a /husky1 isto ispred svega nije potrebno dodat jer to doda namespace-->
    <param name="local_costmap/obstacles_laser/laser/topic" value="$(arg scan_topic)"/>
    <param name="global_costmap/obstacles_laser/laser/topic" value="$(arg scan_topic)"/>
    
    <param name="local_costmap/update_frequency" value="5"/>

    
    <!-- 
    param name="global_costmap/update_frequency" value="6.0"/> inace je 4,, nije nis pomoglo
    -->
  

  </node>

</launch>